src/MongoDB/Driver/Server.cpp

```diff
@@ -138,169 +162,237 @@ Array HHVM_METHOD(MongoDBDriverServer, __debugInfo)
-		return (bool) (sd->type == MONGOC_SERVER_RS_SECONDARY);
+	if ((sd = mongoc_client_get_server_description(data->m_client, data->m_server_id))) {
+		bool isType;
+
+		isType = (strcmp(mongoc_server_description_type(sd), hippo_server_description_type_map[HIPPO_SERVER_RS_SECONDARY].name) == 0);
+
+		mongoc_server_description_destroy(sd);
+
+		return isType;
 	}
 
-	throw MongoDriver::Utils::CreateAndConstruct(MongoDriver::s_MongoDriverExceptionRuntimeException_className, "Failed to get server description, server likely gone: " + HPHP::Variant(error.message).toString(), HPHP::Variant((uint64_t) 0));
+	throw MongoDriver::Utils::CreateAndConstruct(MongoDriver::s_MongoDriverExceptionRuntimeException_className, "Failed to get server description", HPHP::Variant((uint64_t) 0));
 }
 
 bool HHVM_METHOD(MongoDBDriverServer, isArbiter)
 {
 	MongoDBDriverServerData* data = Native::data<MongoDBDriverServerData>(this_);
 	mongoc_server_description_t *sd;
-	bson_error_t error;
 
-	if ((sd = mongoc_topology_description_server_by_id(&data->m_client->topology->description, data->m_server_id, &error))) {
-		return (bool) (sd->type == MONGOC_SERVER_RS_ARBITER);
+	if ((sd = mongoc_client_get_server_description(data->m_client, data->m_server_id))) {
+		bool isType;
+
+		isType = (strcmp(mongoc_server_description_type(sd), hippo_server_description_type_map[HIPPO_SERVER_RS_ARBITER].name) == 0);
+
+		mongoc_server_description_destroy(sd);
+
+		return isType;
 	}
 
-	throw MongoDriver::Utils::CreateAndConstruct(MongoDriver::s_MongoDriverExceptionRuntimeException_className, "Failed to get server description, server likely gone: " + HPHP::Variant(error.message).toString(), HPHP::Variant((uint64_t) 0));
+	throw MongoDriver::Utils::CreateAndConstruct(MongoDriver::s_MongoDriverExceptionRuntimeException_className, "Failed to get server description", HPHP::Variant((uint64_t) 0));
 }
 
 bool HHVM_METHOD(MongoDBDriverServer, isHidden)
 {
 	MongoDBDriverServerData* data = Native::data<MongoDBDriverServerData>(this_);
 	mongoc_server_description_t *sd;
-	bson_error_t error;
 
-	if ((sd = mongoc_topology_description_server_by_id(&data->m_client->topology->description, data->m_server_id, &error))) {
-		bson_iter_t iter;
+	if ((sd = mongoc_client_get_server_description(data->m_client, data->m_server_id))) {
+		const bson_t       *is_master = mongoc_server_description_ismaster(sd);
+
+		Variant v_last_is_master;
+		Array   a_last_is_master;
+
+		hippo_bson_conversion_options_t options = HIPPO_TYPEMAP_DEBUG_INITIALIZER;
+
+		BsonToVariantConverter convertor(bson_get_data(is_master), is_master->len, options);
+		convertor.convert(&v_last_is_master);
+		a_last_is_master = v_last_is_master.toArray();
 
-		return !!(bson_iter_init_find_case(&iter, &sd->last_is_master, "hidden") && bson_iter_as_bool(&iter));
+		mongoc_server_description_destroy(sd);
+
+		return a_last_is_master.exists(s_MongoDriverServer_hidden) && !!a_last_is_master[s_MongoDriverServer_hidden].toBoolean();
```

Comment:
Why not `bson_iter_init_find_case(&iter, mongoc_server_description_ismaster(sd), "hidden") && bson_iter_as_bool(&iter)` here, as you were doing previously (and we still do in PHPC)? It avoids BSON conversion and I reckon it's easier to read, too.